### PR TITLE
Shadownotes during note entry

### PR DESCRIPTION
**Pull request by 19joho66**
When I first used Musescore, I was a bit confused by the display of the notes during note entry. I expected them to be the note I was entering, and not just a filled or unfilled notehead. I also found it awkward, that after klicking the left mouse button in oder to enter a note, the entry cursor moved, but the mousecursor did not.

In this pull request I changed the behaviour to adress these issues. The changes are:

During note entry, shadow notes now also show duration of note
During drum note, entry shadow notes show the distinct noteheads
During note entry, the mouse cursor now follows the input cursor

### CODE REVIEW

**Comment by ericfont** [comment]
@19joho66 I've tested out your PR, and I like it.  It lets me know what duration I've selected, so I don't have to move my eyeballs at the toolbar if I can't remember what duration I had selected.

I believe one reason why your PR hasn't gotten any response is because one of the early steps in the MuseScore Git workflow "Choose something to work on" https://musescore.org/en/developers-handbook/git-workflow#Choose-something-to-work-on is that you check out the issue tracker: https://musescore.org/en/project/issues and so I think it should be implied that after first searching for your feature, you should create an issue there if you don't see it, than other developers can give feedback before your implement it.  (and actually usually before I create an issue in the tracker, I first search and discuss in the https://musescore.org/en/forum/7 which is feature requests forum, so that non-programmer users can give feedback, and so can idenfity exactly what is the feature description will be for when I create the issue.)

Second the instructions for https://musescore.org/en/developers-handbook/git-workflow#Create-a-Topical-Development-Branch say to label the branch after the number the issue tracker assigns, like xxxxx-short-title, but I see you've already used the date.  I can't say if that is acceptable or not becuase it is lasconic who is in charge of accepting PRs...I'm just a part-time contributor.

Third, you forgot the important step in https://musescore.org/en/developers-handbook/git-workflow#Write-some-code! that to begin the commit message with "fix #xxxxx", with xxxx the issue number in the issue tracker, so that the issue will be marked as fixed automatically when your commit is pushed in the main repository.

And at this point, you might need to rebase to current master https://musescore.org/en/developers-handbook/git-workflow#Rebase-your-branch-on-the-latest-upstream

So in summary what I would suggest (1.) make issue in musescore.org tracker, (2.) rebase with current master, (3.) make new commit with message "fix #xxxxx:" and squash as new branch xxxxx-shadow-duration with that *single* commit, then (4.) rebsubmit the PR (maybe has to be new PR if different branch, I don't know???), then (5.) change the satus of your issue in the issue tracker from "active" to "patch (code needs review)".   Then I'm sure lasconic will eventually look at it.

**Comment by lasconic** [comment]
Like Eric said. It would be really helpful to have some feedback from the community. Especially when it comes to modifying the user interface.

**Comment by 19joho66** [comment]
I've introduced this feature in the feature request forum, in order to collect some opinions about this.

**Comment by Jojo-Schmitz** [comment]
for the record, the discussion is at https://musescore.org/en/node/99536

**Comment by ZackTheCardshark** [comment]
@lasconic @19joho66 This was brought up again at https://musescore.org/en/node/112311#comment-506176. I think the opinions that were "collected" were unanimously in favor of this, and there are no merge conflicts.

**Comment by lasconic** [comment]
Sure, give me some more time (yes I know it's almost 6 months...), I will check this one and the other one from @19joho66

**Comment by lasconic** [comment]
There is a bit of work to do before merging this PR. In particular, 
* the drawing of the notes doesn't work for spatium value different from the default. There is a -18 hardcoded.
* I'm not sure sym.cpp needs to have anything for shadow note drawing. It seems that it belongs to shadow notes.
* Coding style is not respected. Use of this, indentation etc...

I started a cleanup but I will be happy to let @19joho66 do it if he is around.

**Comment by lasconic** [comment]
There is also something weird with the mouse note entry. The mouse cursor is moving without the user moving the mouse since it follows the entered notes. I can't reproduce this in master so I assume it has been added with this PR. I'm not sure if we want this behavior,  @IsaacWeiss @ericfont did you experienced this as well? Do you like it?

File: libmscore/sym.cpp [hunk 1]
```diff
@@ -5390,18 +5390,58 @@ void ScoreFont::draw(SymId id, QPainter* painter, qreal mag, const QPointF& pos,
...
+
+      if(ids[1]!=SymId::noSym) {
+            if(ids[1]!=SymId::flagInternalUp) {
+                  pos.ry() -= (sym(ids[1]).bbox().height());
+                  p->drawLine(QLineF(pos.rx(), yOffset, pos.rx(), pos.ry()-yOffset));
+                  pos.rx() -= penWidth/3;
+                  draw(ids[1], p, mag, pos, scale);
+                  }
+            else {
+                  p->drawLine(QLineF(pos.rx(), yOffset, pos.rx(), -18));
```
Comment (lasconic) [inline libmscore/sym.cpp]: the -18 I talked about.

File: mscore/scoreview.cpp [hunk 2]
```diff
@@ -3522,6 +3530,15 @@ void ScoreView::noteEntryButton(QMouseEvent* ev)
       ChordRest* cr = _score->inputState().cr();
       if (cr)
             adjustCanvasPosition(cr, false);
+
+      // get the position of the cursor rect for the next noteentry
```
Comment (lasconic) [inline mscore/scoreview.cpp]: This is probably the code that make the autoadvance work. This is unrelated to the shadown notes, not sure we want to move the mouse cursor if the user doesn't move her mouse...

**Comment by ericfont** [comment]
I don't remember experiencing that mouse behavior...I could try compiling & trying this now.

**Comment by ericfont** [comment]
Looks like lasconic found that code...I would agree shouldn't move the mouse cursor if user doesn't move the mouse.

**Comment by 19joho66** [comment]
As statet in the very first comment (at the end: "During note entry, the mouse cursor now follows the input cursor") this is a designed behavior. That way you don't have to "search" for the next note position but can keep on clicking the notes in. It's like the keyboard input where the input cursor jumps to the next position. This is quite useful when you input a longa or work with several voices. When you use a zoom mode, where you don't see all the measures, this is also a nice feature.

**Comment by 19joho66** [comment]
Sorry, didn't mean to close this

**Comment by lasconic** [comment]
Thank you for your input. I don't know any other software that will move the mouse cursor without the user moving the house. So I would do this in additional PR.

File: libmscore/sym.cpp [hunk 3]
```diff
@@ -5390,18 +5390,58 @@ void ScoreFont::draw(SymId id, QPainter* painter, qreal mag, const QPointF& pos,
...
+void ScoreFont::draw(const std::vector<SymId>& ids, QPainter* p, qreal mag, const QPointF& _pos) const
+      {
+      qreal scale = p->worldTransform().m11();
+      draw(ids, p, mag, _pos, scale);
+      }
+
 void ScoreFont::draw(const std::vector<SymId>& ids, QPainter* p, qreal mag, const QPointF& _pos, qreal scale) const
       {
       QPointF pos(_pos);
+      int x=0;
```
Comment (lasconic) [inline libmscore/sym.cpp]: These modifications destroy the drawing of trills and arpeggios.

**Comment by IsaacWeiss** [comment]
@lasconic, I hadn't noticed the mouse being moved before. I can't say I like it. Trying to enter a chord with the mouse is very difficult like this. 

Also, the trill thing obviously is a serious issue that needs to be fixed.

**Comment by lasconic** [comment]
I fixed most of the issues (arpeggios, small staves etc...). Also decoupled the ShadowNote from Score a bit. Removed useless new class etc... I also removed the mouse move. See https://github.com/musescore/MuseScore/pull/2634

**Comment by thomasbonte** [comment]
@19joho66 Your shadownote contribution got a mention in @IsaacWeiss latest blog post https://musescore.org/en/user/101731/blog/2016/07/01/developing-musescore-3-making-things-easier
